bincrafters_envy/base.py

```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# vim: tabstop=8 expandtab shiftwidth=4 softtabstop=4
+
+import os
+import six
+import fnmatch
+import requests
+import json
+import logging
+from abc import abstractmethod, ABCMeta
+
+
+@six.add_metaclass(ABCMeta)
+class Base(object):
+    def __init__(self, config, host):
+        self._auth = None
+        self._token = self._read_token(config)
+        self._host = self._read_host(config) or host
+        self._headers = dict()
+
+    @abstractmethod
+    def update(self, project_slug, env_vars, encrypted_vars):
+        raise NotImplementedError('"update" method is abstract')
+
+    @abstractmethod
+    def add_one(self, project_slug):
+        raise NotImplementedError('"add_one" method is abstract')
+
+    @abstractmethod
+    def remove_one(self, project_slug):
+        raise NotImplementedError('"remove_one" method is abstract')
+
+    @abstractmethod
+    def list(self):
+        raise NotImplementedError('"list" method is abstact')
+
+    @abstractmethod
+    def exists(self, project_slug):
+        raise NotImplementedError('"exists" method is abstract')
+
+    def _get(self, **kwargs):
+        return self._request(method="GET", **kwargs)
+
+    def _post(self, **kwargs):
+        return self._request(method="POST", **kwargs)
+
+    def _delete(self, **kwargs):
+        return self._request(method="DELETE", **kwargs)
+
+    def _patch(self, **kwargs):
+        return self._request(method="PATCH", **kwargs)
+
+    def _put(self, **kwargs):
+        return self._request(method="PUT", **kwargs)
+
+    def _request(self, method, url, data=None, expected_status=200):
+        url = self._endpoint + url
+        r = requests.request(method=method, url=url, headers=self._headers, auth=self._auth, data=data)
+        if r.status_code != expected_status:
+            raise Exception('%s %s request failed %s %s' % (self.name, method, r.status_code, r.content))
+        content_type = r.headers.get("Content-Type") or ""
+        if "application/json" in content_type:
+            # charmap' codec can't encode character '\u015f' in position 169832: character maps to <undefined>
+            return json.loads(r.content.decode('utf-8', 'replace'))
```

Comment:
```suggestion
            return r.json()
```
this does not work?

Reply:
no idea, need to check, because some json from appveyor contained incorrect utf-8 sequences